XIPs/xip-48-content-type-local-db-integration.md

```diff
@@ -0,0 +1,338 @@
+
+## Backward compatibility
+
+<!-- All XIPs that introduce backward incompatibilities must include a section describing these incompatibilities and their severity. The XIP must explain how the author proposes to deal with these incompatibilities. XIP submissions without a sufficient backward compatibility treatise may be rejected outright. -->
+
+As alluded to above, one downside of moving content types from JSON to protobufs is that it will introduce the possibility of two users on different SDK versions expecting to send and receive messages with different encoding logic. Luckily, the `EncodedContent` wrapper struct that wraps all content types has two fields to help deal with these scenarios: the `ContentTypeId` and the `fallback` string. See Protobuf definitions below:
+
+```proto
+// ContentTypeId is used to identify the type of content stored in a Message.
+message ContentTypeId {
+  string authority_id = 1;  // authority governing this content type
+  string type_id = 2;       // type identifier
+  uint32 version_major = 3; // major version of the type
+  uint32 version_minor = 4; // minor version of the type
+}
+
+// EncodedContent bundles the content with metadata identifying its type
+// and parameters required for correct decoding and presentation of the content.
+message EncodedContent {
+  // content type identifier used to match the payload with
+  // the correct decoding machinery
+  ContentTypeId type = 1;
+  // optional encoding parameters required to correctly decode the content
+  map<string, string> parameters = 2;
+  // optional fallback description of the content that can be used in case
+  // the client cannot decode or render the content
+  optional string fallback = 3;
+  // optional compression; the value indicates algorithm used to
+  // compress the encoded content bytes
+  optional Compression compression = 5;
+  // encoded content itself
+  bytes content = 4;
+}
+```
+
+### Backward compatibility cases to consider
+
+Users on older SDK versions both using JSON encoding will be fine, and new users both using protobuf encoding will be fine. The cases where backward compatibility will be an issue are the following:
+
+#### 1. An older XMTP SDK version receives a message with a new protobuf content type
+
+In this case, your SDK version should recognize that it does not know how to decode the content type because of it's higher ContentTypeId version, and it will instead use the fallback string of the EncodedContent wrapper struct.
+
+#### 2. An older XMTP SDK version sends messages with old JSON content types to users on newer SDK versions
+
+One option is for integrator targeted SDKs to retain code for decoding older JSON content types. A second option is to update integrator targeted SDKs to only support the latest protobuf based content types that will work in new content type based message functions, and to only use fallback text for messages that are not compatible with the latest content types.
```

Comment:
Agree that backwards compatibility isn't the top priority. The only thing that makes me lean towards the first option is how straightforward it should be.

Will make integrators lives easier if they can trust that any reaction they see will be decoded properly. I worry that if we don't integrators will see the fallback from older messages and try to parse it or manually decode the content.

Reply:
> I worry that if we don't integrators will see the fallback from older messages and try to parse it or manually decode the content.

Good point, and I agree it shouldnt be too difficult. Updated here https://github.com/xmtp/XIPs/pull/74/commits/f403869d734a4ff58f1379210a79df34d29d079c. (also simplified the backwards compatability section since it can be summarized as two cases instead of four)